crates/tx-graph/src/pog_musig_functor.rs

```diff
@@ -0,0 +1,328 @@
+    }
+
+    pub fn as_ref(&self) -> PogMusigF<&T> {
+        PogMusigF {
+            challenge: &self.challenge,
+            pre_assert: &self.pre_assert,
+            post_assert: self.post_assert.each_ref(),
+            payout_optimistic: self.payout_optimistic.each_ref(),
+            payout: self.payout.each_ref(),
+            disprove: &self.disprove,
+            slash_stake: self
+                .slash_stake
+                .iter()
+                .map(|x| x.each_ref())
+                .collect::<Vec<[&T; 2]>>(),
+        }
+    }
+
+    pub fn map<U>(self, mut f: impl FnMut(T) -> U) -> PogMusigF<U> {
+        PogMusigF {
+            challenge: f(self.challenge),
+            pre_assert: f(self.pre_assert),
+            post_assert: self.post_assert.map(&mut f),
+            payout_optimistic: self.payout_optimistic.map(&mut f),
+            payout: self.payout.map(&mut f),
+            disprove: f(self.disprove),
+            slash_stake: self
+                .slash_stake
+                .into_iter()
+                .map(|[a, b]| [f(a), f(b)])
+                .collect::<Vec<[U; 2]>>(),
+        }
+    }
+
+    pub fn zip<U>(self, other: PogMusigF<U>) -> PogMusigF<(T, U)> {
+        PogMusigF {
+            challenge: (self.challenge, other.challenge),
+            pre_assert: (self.pre_assert, other.pre_assert),
+            post_assert: self
+                .post_assert
+                .into_iter()
+                .zip(other.post_assert)
+                // TODO(proofofokeags): figure out how to do without intermediate Vec
+                .collect::<Vec<(T, U)>>()
+                .try_into()
+                .ok()
+                .unwrap(),
+            payout_optimistic: self
+                .payout_optimistic
+                .into_iter()
+                .zip(other.payout_optimistic)
+                // TODO(proofofokeags): figure out how to do without intermediate Vec
+                .collect::<Vec<(T, U)>>()
+                .try_into()
+                .ok()
+                .unwrap(),
+            payout: self
+                .payout
+                .into_iter()
+                .zip(other.payout)
+                // TODO(proofofokeags): figure out how to do without intermediate Vec
+                .collect::<Vec<(T, U)>>()
+                .try_into()
+                .ok()
+                .unwrap(),
+            disprove: (self.disprove, other.disprove),
+            slash_stake: self
+                .slash_stake
+                .into_iter()
+                .zip(other.slash_stake)
+                .map(|(a, b)| {
+                    a.into_iter()
+                        .zip(b)
+                        // TODO(proofofokeags): figure out how to do without intermediate Vec
+                        .collect::<Vec<(T, U)>>()
+                        .try_into()
+                        .ok()
+                        .unwrap()
+                })
+                .collect(),
+        }
+    }
```

Comment:
```suggestion 
  pub fn zip<U>(self, other: PogMusigF<U>) -> PogMusigF<(T, U)> {
      PogMusigF {
          challenge: (self.challenge, other.challenge),
          pre_assert: (self.pre_assert, other.pre_assert),
          post_assert: array::from_fn(|i| (self.post_assert[i].clone(), other.post_assert[i].clone())),
          payout_optimistic: array::from_fn(|i| (self.payout_optimistic[i].clone(), other.payout_optimistic[i].clone())),
          payout: array::from_fn(|i| (self.payout[i].clone(), other.payout[i].clone())),
          disprove: (self.disprove, other.disprove),
          slash_stake: self
              .slash_stake
              .into_iter()
              .zip(other.slash_stake)
              .map(|(a, b)| array::from_fn(|i| (a[i].clone(), b[i].clone())))
              .collect(),
      }
  }
```
similar can be done with `zip_apply` below.

Reply:
This doesn't work because it requires introducing Clone constraints on T and U. I tried to keep the assumptions about contents as minimal as possible.

Reply:
Yeah, that's fair